a10networks/a10-octavia
# STACK-1063 Implement VE and VLAN for Hardware Devices (DHCP only)

## Description
- Added feature to enable communication over vlans with traffic isolation for Hardware devices.

## Jira Ticket
[STACK-1063](https://a10networks.atlassian.net/browse/STACK-1063)

## Technical Approach
- Added Config option - `network_type` for user to specify `vlan` or `flat` hierarchy
- Under `[hardware_thunder]`, user can add `interface_vlan_map` to specify configuring ve via dhcp or not. Added column in `vthunders` table and `VThunder` model.
- Added validations for `interface_vlan_map` in `utils.py`
- Separated flat and vlan subflows.
- Added method in `utils` to get network driver.
- Added task to create and delete TagEthernet for both loadbalancer and member.
- Added methods to create, get and delete neutron ports.
- Refactored `config_types.py` 


## Config Changes
<pre>
<b>[a10_global]</b>
network_type = 'vlan'

<b>[hardware_thunder]</b>
devices=[
 {
 "project_id": "fake_uuid",
 "username" : "username",
 "password" : "password",
 "ip_address" : "10.43.12.137",
 "device_name" : "rack_1",
 "interface_vlan_map":{
                 {"1":
                       {"11": {"use_dhcp": True}, 
                         "12":{"use_dhcp": True}
                        }
                 }
       }
 }]
</pre>

## Test Cases
- Updated UTs to handle positive and negative test cases.

## Manual Testing
1) Add 1 VIP and 2 Members in 3 different VLANs and send traffic
2) Delete the Members and VIP one by one and check the VLANs and VE deleted
3) Add VIP and Member from same VLAN and another member in different VLAN send traffic
4) After above case Delete the Member on VIP VLAN and send traffic
5) Change the VE IPs in a10-octavia.conf, restart a10-controller worker and add a member. Check the VE IPs changed accordingly in thunder. Send traffic.
6) Again change VE IPs in conf, restart controller worker and disable a member. Check the VE IPs changed accordingly in thunder. Send traffic.
7) With network_type flat add 1 VIP and 2 Members and send traffic.
8) With above setup remove 1 Member and send traffic.
9) With above setup remove the remaining entities and check configuration.
